wp-includes/mysql/class-wp-mysql-lexer.php

```diff
@@ -0,0 +1,2898 @@
+	 *
+	 * @var int
+	 */
+	private $mysql_version;
+
+	/**
+	 * The SQL modes that should be considered active during tokenization.
+	 *
+	 * This is an integer that represents currently active SQL modes as a bitmask.
+	 * The SQL modes are defined as "SQL_MODE_"-prefixed constants in this class.
+	 * The list of the SQL modes isn't exhaustive, as only some affect tokenization.
+	 *
+	 * @var int
+	 */
+	private $sql_modes;
+
+	/**
+	 * How many bytes from the original SQL payload have been read and tokenized.
+	 *
+	 * This is an internal cursor that is used to track the current position in
+	 * the SQL payload during tokenization. When used as an index in the SQL
+	 * payload, it points to the next byte to read.
+	 *
+	 * @var int
+	 */
+	private $bytes_already_read = 0;
+
+	/**
+	 * Byte offset in the SQL payload where current token starts.
+	 *
+	 * This is used to extract the token bytes after the token is processed.
+	 * The bytes of the current token are represented by "$this->sql" in range
+	 * from "$this->token_starts_at" to "$this->bytes_already_read - 1".
+	 *
+	 * @var int
+	 */
+	private $token_starts_at = 0;
+
+	/**
+	 * Whether the tokenizer is inside an active MySQL-specific comment.
+	 *
+	 * MySQL supports a special comment syntax whose content is recognized as
+	 * a comment by most database engines, but can be treated as SQL by MySQL:
+	 *
+	 *  1. /*! ...  - The content is treated as SQL.
+	 *  2. /*!12345 - The content is treated as SQL when "MySQL version >= 12345".
+	 *
+	 * @var bool
+	 */
+	private $in_mysql_comment = false;
+
+	/**
+	 * @param string $sql The SQL payload to tokenize.
+	 * @param int $mysql_version The version of the MySQL server that the SQL payload is intended for.
+	 * @param int $sql_modes The SQL modes that should be considered active during tokenization.
+	 */
+	public function __construct(
+		string $sql,
+		int $mysql_version = 80038,
+		int $sql_modes = 0
+	) {
+		$this->sql           = $sql;
+		$this->mysql_version = $mysql_version;
+		$this->sql_modes     = $sql_modes;
+	}
+
+	/**
+	 * Read the next token from the SQL payload and return it as a token object.
+	 *
+	 * This method reads bytes from the SQL payload until a token is recognized.
+	 * It starts from "$this->sql[ $this->bytes_already_read ]", advances the
+	 * number of bytes read, and returns a WP_MySQL_Token object. When the end of
+	 * the SQL payload is reached, the method always returns an EOF token.
+	 *
+	 * @return WP_MySQL_Token A token object representing the next recognized token.
+	 */
+	public function next_token(): WP_MySQL_Token {
+		do {
+			$this->token_starts_at = $this->bytes_already_read;
+			$type                  = $this->read_next_token();
+		} while (
+			in_array(
+				$type,
+				array(
+					self::WHITESPACE,
+					self::COMMENT,
+					self::MYSQL_COMMENT_START,
+					self::MYSQL_COMMENT_END,
+				),
+				true
+			)
+		);
+		return new WP_MySQL_Token( $type, $this->get_current_token_bytes() );
```

Comment:
If this is called multiple times after the tokenizing is finished, would it return an EOF token each time? What do you think about returning `false` when there's nothing else to consume?

I also had a brief thought about returning `true` on success as that would make it consistent with the way `WP_HTML_Tag_Processor` works, but I don't have strong opinions here. It seems that would only require an additional `get_token()` method without further benefits.

In case of HTML processor, we also have setters and we can also inspect a bunch of metadata about the current token directly on the processor instance. There's no token class in there because the parser only moves forward. In here, the parser backtracks a lot so it makes sense to reuse the tokenization results.

Reply:
@adamziel Yeah, I agree it's strange to return `EOF` forever.

What about returning `null` when there are no more tokens?

Hm, but what about invalid input? Now it uses the `INVALID_INPUT` token, but we could also stop as soon as an error occurs, and reflect that in the return value. Such an input wouldn't be then passed on to the parser at all. Or, are there use cases, where it would be useful to keep an `INVALID_INPUT` token and try to continue the lexing process?

Then we could consider some of these:
1. Return `WP_MySQL_Token|null`, throw exception on invalid input.
2. Return `WP_MySQL_Token|null|false` (`null` for no more tokens, `false` for an error).
3. Multiple methods (the `next_token()` & `get_token()` scenario).

Reply:
@adamziel
So, do you think something like `WP_MySQL_Token|null|false` for now?

What would "not enough input" mean in this case, and what would be the use case? The lexer just reads the input and creates tokens. When it's an invalid token, it creates an `INVALID_TOKEN` type, and when there is no more input, it's `EOF`. I'm not sure if the incompleteness is deterministic here, like in the case of an unclosed HTML tag. Isn't that more of a parser-level feature? I think that `WP_XML_Processor` is different in the way that it's also a parser at the same time.

As for the `INVALID_TOKEN`, I think it makes sense to never output anything like that and stop at the first error. The `INVALID_TOKEN` approach would decode `b'01xyz'` as `INVALID_INPUT`, `IDENTIFIER`, `INVALID_INPUT`, `EOF`, which doesn't really make much sense and is rather random.

Reply:
> Isn't that more of a parser-level feature? I think that WP_XML_Processor is different in the way that it's also a parser at the same time.

Not at all! At first we may only have `SELECT 1`. If we tokenize that without waiting for more input, we'll get everything mixed up by the time we learn it's actually `SELECT 123, post_content FROM wp_posts`.

Reply:
@adamziel Ah, right, thank you for the detailed explanation! It didn't occur to me that if we need to chunk a big file, we have no idea how to chunk it before we actually tokenize it.

Looking at the code example, I have one question. Is there an advantage to iterating the chunks from outside the lexer, rather than passing a stream into it? In other words, do we need to expose `is_paused...` and other methods? E.g., is it better because it's a more generic API, and a stream would be just one possible use case?

In this case, I agree that we should probably go with the `next_token()` and `get_token()` API.

Reply:
I've tried using PHP streams as the primary API for this kind of operations in AsyncHttp\Client and I failed miserably. I ended up removing that in https://github.com/WordPress/blueprints-library/pull/113.

Tl;dr – every stream wrapper (https:// vs file:// vs memory://) handles nuances differently. Some won't work with stream filters or buckets under specific circumstances or in specific PHP versions and there's almost zero documentation on these very deep use-cases. Unfortunately it's not worth it.

Reply:
I also explored a generic stream interface in https://github.com/adamziel/wxr-normalize/pull/1 and a compelling use-case is yet to emerge.

Reply:
@adamziel Oh, wow, very interesting! I had no idea PHP streams are such a pain. A generic interface sounds great, but looking at the explorations, it also looks pretty complex, so I guess at this stage, we need to work with the `WP_XML_Processor`-like APIs in mind.

I pushed a commit for the `next_token()` and `get_token()` combo. Now, the `remaining_tokens` method makes a little bit less sense, I guess, sided with the next-token/get-token API, but for the moment I kept it in. Let me know what do you think.